internal/pubsub/rabbitpubsub/rabbit.go

```diff
@@ -0,0 +1,179 @@
+		return nil, err
+	}
+	return pubsub.NewTopic(t), nil
+
+}
+
+func newTopic(conn *amqp.Connection, name string) (*topic, error) {
+	t := &topic{
+		exchange: name,
+		conn:     conn,
+	}
+	if err := t.establishChannel(); err != nil {
+		return nil, err
+	}
+	return t, nil
+}
+
+// establishChannel creates an AMQP channel if necessary. According to the amqp
+// package docs, once an error is returned from the channel, it must be discarded and
+// a new one created.
+//
+// Must be called with the t.mu held.
+func (t *topic) establishChannel() error {
+	// TODO(jba): support context.Context
+	if t.ch != nil {
+		// We already have a channel; nothing to do.
+		return nil
+	}
+	// Create a new channel.
+	ch, err := t.conn.Channel()
+	if err != nil {
+		return err
+	}
+	// Put the channel into a mode where confirmations are delivered for each publish.
+	if err := ch.Confirm(wait); err != nil {
+		return err
+	}
+	t.ch = ch
+	// Get a Go channel which will hold acks from the server. The server
+	// will send an ack for each published message to confirm that it was received.
+	t.pubc = ch.NotifyPublish(make(chan amqp.Confirmation)) // NotifyPublish returns its arg
+	return nil
+}
+
+// SendBatch implements driver.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, ms []*driver.Message) error {
+	// It is simplest to allow only one SendBatch at a time. Allowing concurrent
+	// calls to SendBatch would complicate the logic of receiving publish
+	// confirmations and returns. We can go that route if performance warrants it.
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if err := t.establishChannel(); err != nil {
+		return err
+	}
+
+	// Read from the channel established with NotifyPublish.
+	// Do so concurrently or we will be deadlock with the Publish RPC.
+	// (The amqp package docs recommend setting the capacity of the channel
+	// to the number of messages to be published, but we can't do that because
+	// we want to reuse the channel for all calls to SendBatch.)
+	errc := make(chan error, 1)
+	go func() {
+		// Consume all the acknowledgments for the messages we are publishing.
+		// Since this method holds the lock, we expect exactly as many acks as messages.
+		//
+		// This goroutine can safely access the mutex-protected fields t.ch and t.pubc
+		// because its lifetime is within the lifetime of the lock held by SendBatch.
+		//
+		// TODO(jba): look at AMQP "returns" for more information about publish errors.
+		// See https://godoc.org/github.com/streadway/amqp#Channel.NotifyReturn.
+		ok := true
+		for range ms {
+			select {
+			case <-ctx.Done():
+				errc <- ctx.Err()
+				return
+			case conf, ok := <-t.pubc:
+				if !ok {
+					// t.pubc was closed
+					errc <- errors.New("rabbitpubsub: publish listener closed unexpectedly")
+					t.ch = nil // re-create the channel on next use
+					return
+				}
+				if !conf.Ack {
+					ok = false
+				}
+			}
+		}
+		if !ok {
+			errc <- errors.New("rabbitpubsub: ack failed on publish")
```

Comment:
What happens if some sends were acked and others were not? Does `driver` say anything about this?

Reply:
I think it should be an error if even one message isn't sent successfully, which is what I do here.

I think we should work on the doc for driver.SendBatch. Currently it reads
```
// Only one RPC should be made to send the messages, and the returned                                                                                                                                                                     
// error should be based on the result of that RPC. Implementations                                                                                                                                                                       
// that send only one message at a time should return a non-nil error                                                                                                                                                                     
// if len(ms) != 1.  
```
But I don't think the concrete impl does anything to make sure `len(ms) ==  1`.

Reply:
I think that should say that ideally only one RPC is necessary, but more may be, and len(ms) may end up being > 1.

Reply:
@ijt

Reply:
I think this is a design mistake: "Implementations                                                                                                                                                           that send only one message at a time should return a non-nil error                                                                                                                                                                     
if len(ms) != 1."

A better approach would be to have a method like `driver.Topic.MaxBatchSize()` that says how many messages can go in a batch.

Reply:
I filed #841 for this discussion.